opm/grid/cpgpreprocess/preprocess.c

```diff
@@ -716,29 +717,178 @@ vert_size(const struct grdecl *in,
+
+
+/* ---------------------------------------------------------------------- */
+static enum CoordinateSystemType
+classify_geometry(const double triple)
+/* ---------------------------------------------------------------------- */
+{
+    if (triple > 0.0) {
+        return RightHanded;
+    }
+    else if (triple < 0.0) {
+        return LeftHanded;
+    }
+    else {
+        return Inconclusive;
+    }
+}
+
+
+/* ---------------------------------------------------------------------- */
+static enum CoordinateSystemType
+get_cell_type(const struct grdecl *in,
+              const size_t         i,
+              const size_t         j,
+              const size_t         k,
+              const double         sign,
+              const size_t         off[8])
+/* ---------------------------------------------------------------------- */
+{
+    const size_t p0 = i + j*(in->dims[0] + 1);
+    const size_t pi = p0 + 1;
+    const size_t pj = p0 + in->dims[0] + 1;
+    const size_t io = 2*i + 2*in->dims[0]*(2*j + 2*in->dims[1]*2*k);
+
+    double triple, origin[3], I[3], J[3], K[3];
+
+    vertex_coord(in, p0, in->zcorn[io + off[0]], origin);
+    vertex_coord(in, pi, in->zcorn[io + off[1]], I);
+    vertex_coord(in, pj, in->zcorn[io + off[2]], J);
+    vertex_coord(in, p0, in->zcorn[io + off[4]], K);
+
+    triple = sign * bounding_box_triple_product(origin, I, J, K);
+
+    return classify_geometry(triple);
+}
+
+
+/* ---------------------------------------------------------------------- */
+static enum CoordinateSystemType
+coodinate_system_type_cell_criterion(const struct grdecl *in,
+                                     const double         sign,
+                                     const size_t         off[8])
+/* ---------------------------------------------------------------------- */
+{
+    size_t i, nx, j, ny, k, nz, c;
+    size_t ctype[3] = {0};
+
+    nx = in->dims[0];
+    ny = in->dims[1];
+    nz = in->dims[2];
+
+    if (in->actnum == NULL) {
+        for (k = 0, c = 0; k < nz; k++) {
+            for (j = 0; j < ny; j++) {
+                for (i = 0; i < nx; i++, c++) {
+                    if (vert_size(in, c, off) > 0.0) {
+                        ++ ctype[get_cell_type(in, i, j, k, sign, off)];
+                    }
+                }
+            }
+        }
+    }
+    else {
+        for (k = 0, c = 0; k < nz; k++) {
+            for (j = 0; j < ny; j++) {
+                for (i = 0; i < nx; i++, c++) {
+                    if ((in->actnum[c] != 0) && (vert_size(in, c, off) > 0.0)) {
+                        ++ ctype[get_cell_type(in, i, j, k, sign, off)];
+                    }
+                }
+            }
+        }
+    }
+
+    if ((ctype[Inconclusive] > ctype[RightHanded]) &&
+        (ctype[Inconclusive] > ctype[LeftHanded]))
+    {
+        return Inconclusive;
+    }
+    else if (ctype[RightHanded] > ctype[LeftHanded]) {
+        return RightHanded;
+    }
+    else {
+        return LeftHanded;
+    }
```

Comment:
Totally unclear to me. So it is ok to have different cells as long as there are few inconclusive one. And if only there is one more right-handed cell than left-handed ones everything is assumed to be right-handed?

Reply:
> So it is ok to have different cells as long as there are few inconclusive one. And if only there is one more right-handed cell than left-handed ones everything is assumed to be right-handed?

All very good points, and I hinted at this in an [earlier comment](https://github.com/OPM/opm-grid/pull/625#issuecomment-1415686665).  It's mostly based on a "likelihood"/empirical argument, but there's no theory behind it.  It might be better, but probably more expensive, to try to form a model bounding box for the columns with active cells only.  For now I think this is "good enough", but I am quite prepared to be shown otherwise.